segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/BookKeeperLog.java

```diff
@@ -115,12 +121,10 @@
         this.executorService = Preconditions.checkNotNull(executorService, "executorService");
         this.closed = new AtomicBoolean();
         this.logNodePath = HierarchyUtils.getPath(logId, this.config.getZkHierarchyDepth());
-        this.lastAppendAddress = new AtomicReference<>(new LedgerAddress(0, 0));
         this.traceObjectId = String.format("Log[%d]", logId);
-        this.rolloverInProgress = new AtomicBoolean();
-        this.retryPolicy = config.getRetryPolicy()
-                                 .retryWhen(BookKeeperLog::isRetryable)
-                                 .throwingOn(Exception.class);
+        this.writes = new ArrayDeque<>();
+        this.writeProcessor = new SequentialAsyncProcessor(this::processWritesSync, this.executorService);
+        this.rolloverProcessor = new SequentialAsyncProcessor(this::rollover, this.executorService);
```

Comment:
They have different jobs and can execute in parallel. During rollover, we create a new ledger, and then update metadata and close the existing ledger (and swap pointers). Creating a new ledger takes some time (500ms or so), and there is no reason to stop processing appends while that happens. As such we keep processing appends while we are concurrently working on rolling over. 

When we close the old ledger, the in-flight writes to that one will fail with LedgerClosedException, and we will detect a new ledger is in place at which point we retry those writes on the new ledger. The WriteProcessor is designed to not start writes on a new ledger until all writes to the previous one are done (otherwise we could have out-of-order writes).